### ISSUE TITLE
[Techstack-beta] Design: Narrow menu button accessibility problems

### ISSUE THREAD

**Issue by samueldr** [NixOS/nixos-homepage#1288]
Follow-up to the incorrect/incomplete fix for #1253.

* * *

In web development, generally speaking, `tabindex` shouldn't be used, and *if* it is needed, it generally means that it's used to cover another bug of deficiency in the design.

> [!WARNING]  
> Warning: You are recommended to only use `0` and `-1` as tabindex values. Avoid using `tabindex` values greater than `0` and CSS properties that can change the order of focusable HTML elements ([Ordering flex items](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_flexible_box_layout/Ordering_flex_items)). Doing so makes it difficult for people who rely on using keyboard for navigation or assistive technology to navigate and operate page content. Instead, write the document with the elements in a logical sequence.
— https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/tabindex

So, already, `tabindex=2` (and 1 for the logo) here should be considered unexpected as it makes keyboard navigation happen in an order different from visual/expected.

* * *

Using Firefox's Accessibility developer tool we see:

![image](https://github.com/NixOS/nixos-homepage/assets/132835/7706e0f9-4740-471b-8491-25f2a9bf36c1)

The warning links to:

 - https://developer.mozilla.org/en-US/docs/Web/Accessibility/Understanding_WCAG/Keyboard#avoid_using_tabindex_attribute_greater_than_zero

* * *

Let's move to Chromium with the Vimium extension. Vimium is a *really amazing* tool to find what I call **accessibility smells**. In itself it is not exactly (but also is) an accessibility tool, but the way it combs through the page to find element to navigate to is still representative of how different tools see the site.

To reduce noise, I use the Explore page, which is relatively simple.

Pressing `f` (follow link / activate item) one can see that on the new implementation the element is not seen as one that can be activated.

![image](https://github.com/NixOS/nixos-homepage/assets/132835/d13152f7-8eda-4307-88d7-46b89ca9a653)

* * *

Still in Chromium, its Accessibility developer tool differs greatly from Firefox's. (And I kinda prefer its presentation when dealing with a site.)

Comparing the *Full-Page Accessibility Tree* (ignoring anything outside the banner, and focusing on the narrow menu navigation), we can see that there is no mention of the widget to toggle the menu in the new implementation.

![image](https://github.com/NixOS/nixos-homepage/assets/132835/0084423c-bb2c-4c20-9853-8246508d32c3)

So as far as tools using that model of the accessibility tree, the menu does not exist in a narrow view.

* * *

Quoting from [SCR35](https://www.w3.org/WAI/WCAG21/Techniques/client-side-script/SCR35.html):

>  In order to ensure that scripted actions can be invoked from the keyboard, they are associated with "natively actionable" HTML elements (links and buttons). The `onclick` event of these elements is device independent. While "`onclick`" sounds like it is tied to the mouse, the `onclick` event is actually mapped to the default action of a link or button. The default action occurs when the user clicks the element with a mouse, but it also occurs when the user focuses the element and hits enter or space, and when the element is triggered via the accessibility API.

In this instance, the fix to make the thing keyboard-accessible is:

 - Use no `tabindex` at all
 - Removing the keyboard handling (which is sometimes incomplete since it only handles *Enter*).
 - Replacing the generic semantics-free DOM element (the `<div>`) by, preferably, a `<button>`.

(`<a>` should be reserved for links that do page navigation, ensuring none of its subtle semantics are left, if it needs to be customized.)

Since `onClick` is used (via the `click` event listener) the *implicit and intrinsic* behaviour of the `<button>` makes it reachable via tab navigation, and can be activated via any user-agent method of activating the button, without having to consider every possible case.
